### ISSUE TITLE
Add support for different signing and encryption keys for SP

### ISSUE THREAD

**Issue by dbollaer** [simplesamlphp/simplesamlphp#1001]
Hello,
I need to work with a third party and they require that I provide SAML metadata with different certificates for encryption and signing (SP). They even require the different certificates to be signed by a CA.

You already provide support for parsing metadata which has this behaviour. But I don't this it can be reused to add it to the authsource.

I would like support in the SP authsource to add specific certificates and keys for signing and encryption.

If I missed something in the documentation, or there is a module which enables this. Could you please send me a link?

Thank you in advance.
Danny

**Comment by dbollaer** [comment]
If this a low priority feature. Meaning you want it, but not right now, I can look at implementing this feature myself.
It would be a great help if you can give me some advice on the following aspects:

Naming guideline.
for example.

```
signing.certificate
signing.privatekey
signing.privatekey_pass

encryption.certificate
encryption.privatekey
encryption.privatekey_pass
````

Which classes should I take a look at?
- Generating metadata
- Handling SAML

**Comment by trscavo** [comment]
> Regarding the naming convention, I think you only need to add the encryption-ones to just overrule the current 'coupled' behaviour. I'd also call it `decryption` instead of `encryption`, because that's eventually what we're using it for..

Technically the certificate is used for encryption and the private key is used for decryption. Even so, I would call the latter `encryption.privatekey` but YMMV.

**Comment by tvdijen** [comment]
Nope, unless I'm horribly wrong here, the certificate is used only for publication in _our metadata_.. Encryption is done with the IDP's certificate from it's metadata...

**Comment by dbollaer** [comment]
While looking at the code from key roll over, documentation and your feedback, I was wondering if can emulate the behaviour of using different certificates for signing and encryption with the key rollover feature and manually editing the metadata.

steps:
- I enable key-rollover
- I show only the new key/certificate for encryption.
- I show only the old key/certificate for signing.

Then I would have working configuration with different key and certificate to start testing.

If key-rollover is not supported by the third party, is there a reason why this would be a bad idea?
For me it feels incorrect, but I can not lay my finger on it.

Could you please provide feedback?
Thank you in advance.

**Comment by tvdijen** [comment]
I think that should work.. The only thing is that you will have a permanent performance loss, but as long as you don't use 8K keys or something you should be fine.. Also make sure to tell your IDP to never use the generated metadata.

**Comment by msc1978** [comment]
@dbollaer i have the same requirement.. may I ask is this for eidas?

**Comment by dbollaer** [comment]
@msc1978 it is indeed related. But the requirement has been dropped. So at the moment, I will not develop this feature. 
Thank you for help @tvdijen.
Do I close this issue? Or do you plan on adding it later?

**Comment by tvdijen** [comment]
I think it makes sense to support this, so let's leave it open for now..

**Comment by JohnMaguire** [comment]
This blog post says SSP cannot use per-SP signing keys: https://www.stackallocated.com/blog/2020/saml-idp-no-shared-keys/

I am quite sure that this can be done via `saml20-sp-remote` metadata `signing.certificate` and `signing.privatekey`. Maybe the title for this issue should be updated to reflect that?

**Comment by robvanderboom** [comment]
any news on this issue?
for testing i also this for testing idp (adsf) different signing and encryption

**Comment by tvdijen** [comment]
No news
